pytorch_lightning/trainer/supporters.py

```diff
@@ -508,3 +508,25 @@ def _nested_calc_num_data(data: Union[Mapping, Sequence], compute_func: Callable
             new_data.append(x)
 
     return compute_func(new_data)
+
+
+def prefetch_iterator(iterable: Iterable) -> Generator[Tuple[Any, bool], None, None]:
+    """
+    Returns an iterator that pre-fetches and caches the next item.
+    The values are passed through from the given iterable with an added boolean indicating if this is the last item.
+    See `https://stackoverflow.com/a/1630350 <https://stackoverflow.com/a/1630350>`_
+    """
+    it = iter(iterable)
+
+    try:
+        # the iterator may be empty from the beginning
+        last = next(it)
```

Comment:
How about using `last = next(it, None)`, and if None, return?

Reply:
I didn't know you can do that. But isn't this a trouble when the iterator actually returns None as the real element? Then we can't differentiate

Reply:
yeah, that's the concern I had as well.